eggtimer.js

```diff
@@ -1,326 +1,681 @@
+}
+
+function mergingTag(prNum) {
+    return "tags/" + MergingTag + prNum;
+}
+
+function checkMergePreconditions(prContext) {
+    let reviewsParams = commonParams();
+    reviewsParams.number = prContext.pr.number;
+    let reviewsPromise = getReviews(reviewsParams);
+
+    let statusParams = commonParams();
+    statusParams.ref = prContext.pr.head.sha;
+    let statusPromise = getStatuses(statusParams);
+
+    let mergeableParams = commonParams();
+    mergeableParams.number = prContext.pr.number;
+    let mergeablePromise = getPR(mergeableParams);
+
+    let labelsParams = commonParams();
+    labelsParams.number = prContext.pr.number;
+    let labelsPromise = getLabels(labelsParams);
+
+    Promise.all([reviewsPromise, statusPromise, mergeablePromise, labelsPromise]).then((results) => {
+            if (prContext.signaled)
+                return Promise.reject("signaled");
+
+            let reviews = results[0];
+            let statuses = results[1];
+            let mergeable = results[2].mergeable;
+            let labels = results[3];
+
+            if ((mergeable !== true) || !approved(reviews) || !allChecksSuccessful(statuses))
+                return Promise.reject("not ready for merging yet");
+
+            if (markedAsMerged(labels))
+                return Promise.reject("already merged");
+
+            if (markedAsFailed(labels))
+                console.log("PR" + prContext.pr.number + " previous merge was unsuccessful");
+
+            checkBeforeStartMerging(prContext);
+            return Promise.resolve(true);
+    })
+    .catch((err) => {
+        console.error("Can't merge PR" + prContext.pr.number + ":", err);
+        processNextPR();
+    });
+}
+
+function checkBeforeStartMerging(prContext) {
+   let refParams = commonParams();
+   refParams.ref = mergingTag(prContext.pr.number);
+   let tagTreeSha = null;
+   prContext.tagSha = null;
+   console.log("Checking whether to skip due to a previous unsuccessful merge for PR" + prContext.pr.number);
+   getReference(refParams)
+       .then( (sha) => {
+           let params = commonParams();
+           params.sha = sha;
+           prContext.tagSha = sha;
+           let commitPromise = getCommit(params);
+
+           let prParams = commonParams();
+           prParams.number = prContext.pr.number;
+           let prPromise = getPR(prParams);
+
+           return Promise.all([commitPromise, prPromise]);
+       })
+       .then((results) => {
+           let commitObj = results[0];
+           let pr = results[1];
+           assert(pr.number = prContext.pr.number);
+           let params = commonParams();
+           tagTreeSha = commitObj.treeSha;
+           params.ref = "pull/" + prContext.pr.number + "/merge";
+           return getReference(params);
+       })
+       .then((sha) => {
+           let params = commonParams();
+           params.sha = sha;
+           return getCommit(params);
+       })
+       .then((obj) => {
```

Comment:
I was thinking about these results naming, but renaming obj->commitObj does not give any valuable information, because:
1. Evidently, in such 'then' we are getting the result of the last promisificated method (getCommit) in our case, i.e., this 'obj' is its result.
2. Since bfc2af0 almost all promisificated methods return a 'hash object'; we should look at it anyway in order to use it.